### ISSUE TITLE
Improve / unify error handling on Less.js API calls with lessc / browser

### ISSUE THREAD

**Issue by kirillrogovoy** [less/less.js#2951]
Here's the thing: when you use `lessc` from the command line, you always get detailed error message  when parser fails (specifically, error title, filename, line, col). Which is great.

However, when you use `less.render` programmatically, all you see when error occurs is similar to this:

```
Error: Unrecognised input
```

Even if `options.filename` is set, all you see is just an error's title **without** `filename`, `column`, `line`.

Of course, you have access to that information via error properties (since the error is actually is an instance of [`LessError`](https://github.com/less/less.js/blob/3.x/lib/less/less-error.js)), but it's really uncomfortable to manually fetch those properties when I'm catching the error.

All I want is to always have file details right in `err.message` if `options.filename` was set. Not just in additional properties, but in the general human-readable message I see in console when my code fails.

It's actually just a good JS pattern to have `err.message` as detailed as possible.

I'm ready to make these changes to code, however, it seems there are lots of tests to be corrected and I'd glad to hear contributors' opinions before starting.

**Comment by matthew-dean** [comment]
This actually came up for me recently, in very much the same way. I, too, noticed that the way errors are surfaced is not only not detailed, it's also not consistent.

The way I would expect it to work is to have errors surfaced in the most appropriate way for the environment that Less.js is run in. But.... that doesn't actually happen. Errors are produced at different stages, but they're not necessarily surfaced at all, especially when directly calling `parse()` or `render()` methods. Errors are _returned_, but not written into STDERR or into a console.error in the browser.

I ran into this because I was working on simplifying browser tests. To do so, I was having some tests use the `<link />` auto-compile method, and some tests I've been re-writing using a call to `render()` directly in the browser. (This prevents having to load as many pages in Sauce Labs, for example, so the tests can run faster.)

However, Less.js doesn't actually throw errors in the browser when calling those methods. It only raises errors through the special auto-loading in-browser method, which only runs once.

Simarly, in Node.js, Less writes formatted errors to the console not via the core library, but as a special feature of lessc.

(This isn't the only instance where lessc and less-browser mode duplicate each other's work in a way that should be centralized.)

As a secondary problem, Less has a logger, which you can add custom loggers to. And this includes error logging. But Less-generated warnings/errors are not automatically added to the log. Each environment manually both handles errors and separately logs them, and _separately displays them_. Meaning, again, if you call the API, all 3 of these steps can get skipped.

So, it's awesome if you're ready to make changes to this code! (As it's currently a bit of a mess.) I think these are the 3 areas to start pulling Less together.

1. Make error bubbling consistent across Node.js and the browser.
2. Make error display consistent between the Node.js API and lessc.
3. Make error display consistent between the Less API in browser and <link /> processing errors.
4. Look at unifying the Less "logger" with error and warning generation. Theoretically, if an error or warning is created, should it not be automatically added to the log? That makes sense to me.

**Comment by flashhhh** [comment]
Thanks for your reply!

Looks like I hit the tip of the iceberg.

Ok, I'm going to take a closer look at the codebase at whole and then will focus on error messages consistency.

I didn't use **lesscss** in browser at all, but, as far as I understand, this one:

> Automatically display errors without a manual formatting process.

breaks current API since functions should start throwing errors instead of returning them?
Once again, I'm not sure I'm aware of what I'm talking about here. :)

To sum it up, for now I'm going to do following:

1. Explore the codebase.
2. Make core library compose the best errors possible.
3. Use the previous improvement in `lessc` and browser-specific code.
4. Clean up the mess around.
5. Pull request it.

Is it OK to use this issue to ask for any advises from the core team during the process?

**Comment by matthew-dean** [comment]
> breaks current API since functions should start throwing errors instead of returning them

No, I wasn't suggesting throwing them, but displaying them in the console. Unless you think there's a better idea, or that's a bad idea in general? Perhaps calling an API with an error return implies that it should be handled by the user. So, alternatively, the logger could be more "baked in", and referenceable. 

All I know is that the way it's currently structured was, a) confusing, b) inconsistent between environments, c) not what I expected.

In other words, any improvement you do here that makes errors more accessible / friendly will likely be pulled in, lol. So it doesn't need to happen exactly as I laid it out. When someone is willing to do the work on a community project, the solution is somewhat at their discretion. Do what you think makes the most sense.

As an additional note, Less commits have been a little light on inline code documentation. I've been trying to add more JSDoc-style comments with my commits so that each contributor doesn't have to go through the cognitive burden of figuring out what's happening and why.

So, if you could add comments with your commits, as well as adding documentation to existing code that you're working with, I know that would be greatly appreciated by me and other Less.js contributors.

Thanks!

**Comment by matthew-dean** [comment]
Also:

> Is it OK to use this issue to ask for any advises from the core team during the process?

Absolutely! You can also start a pull request and just note that it's work-in-progress, and then do line comments to ask questions. That's easily managed in Github. Then you can address those questions and add the pull request until it's finished.

**Comment by flashhhh** [comment]
Seems like I didn't find any time to start it. I'm going to give it a try soon.
Do I understand it right that I should use the `3.x` branch as a basis? Is there any information except `CONTRIBUTING.md` that would help me to get started?

**Comment by matthew-dean** [comment]
> Do I understand it right that I should use the 3.x branch as a basis?

Yep.

> Is there any information except CONTRIBUTING.md that would help me to get started?

You can always hit up one of the core team people.

**Comment by flashhhh** [comment]
Implemented the solution to my initial concern here: https://github.com/less/less.js/pull/2975. Ready to go on on further issues after it gets reviewed.

**Comment by rjgotten** [comment]
@matthew-dean 
I think there's one crucial point that current error handling is missing and which is also missing from the proposals listed above: compilation errors are not program errors. They are *not* supposed to be subclassed from `Error` or be thrown.

Compilation errors should just be reported back to the API consumer. Whether that is custom code calling `less.render`; the lessc CLI; or the browser.

The better method to handle this is to have one central logging endpoint which catches messages related to compilation errors and then expose that as part of the public API. The CLI wrapper code can redirect these to STDERR; the browser API can redirect these to `console.error` or `console.warn`; and custom code can do whatever it wants.

For actual *program* errors (or errors in plugins), you should have a single top-level try-catch in the CLI wrapper code and in the browser API code which formats the error appropriately and sends it into the same logging end-point.

Crucially, you should *not* catch anything directly from `less.render` because custom code should be able to handle its own exceptions.


------

Note though that plugin functions being executed and resulting in a program error in the function are kind of special, since they're - in a way - a compilation error. They should probably have their own try-catch wrappings where the catch clause sends the error into the logging end-point and then gracefully halts compilation, rather than bubbling the error through to the top.

**Comment by matthew-dean** [comment]
> Note though that plugin functions being executed and resulting in a program error in the function are kind of special, since they're - in a way - a compilation error. They should probably have their own try-catch wrappings where the catch clause sends the error into the logging end-point and then gracefully halts compilation, rather than bubbling the error through to the top.

I've attempted to do something like that in my latest 3.x commits.... although I didn't think about custom function calls. Those should probably be wrapped into try/catch. I did add tests and code to catch:

- An error parsing the plugin (just like you would get a Less error parsing a file)
- An error with the `use` (pre-eval) or `eval` calls.

Because the plugin entry point is an anonymous function, it's non-standard as to how to get the line/column of the error, but I added a regex to get at least the Node case from the error stack (and also Firefox in-browser). Then, because the existing error handling always required a line / column if a file was referenced, I changed error output to optionally output line/column if, of course, they had actual values.

As to everything else, I believe I'm agreement with what you said. I'll add try/catch functions to the list for 3.0.
